### DISCUSSION TITLE
Why isn't rollback the standard behaviour of failing tests in a DAG?

### DISCUSSION THREAD

**Discussion by volvocarsmatt** [apache/airflow#26867] [Q&A] [answered]
Hi!

I'm new to Airflow and with my previous orchestrator I had the practice or issuing a rollback if the data tests in in transaction showed that the update was incorrect. If I'm not misunderstanding, this is not the way Airflow is designed?

Consider for example that I want to insert (or ideally upsert) values into my table with the PostgresOperator. If I can detect that this upsert will violate some sanity checks with SQLColumnCheckOperator or SQLCheckOperator, I would ideally rollback the the insert and fix the new data. Instead I now also have to find a clever way to fix the data in the table or restore the table to its previous state. This is because tasks [should be treated as equivalent to transactions in a database](https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html?highlight=transaction#creating-a-task) in Airflow and a completed transaction commits the change.

While it will protect the downstream of this checks, wouldn't it be better if it also rolled back the change? Have I misunderstood something?

**Comment by potiuk** [comment, marked as answer]
Depends how your operator is written. Airflow does not make any assumption on what happens with DB operation - it is purely implementation detail of particular operator. From the quick look at the implementation - if you use PostgresOperator (and actually PostgresHook underneath) and do not set auto-commit and won't commit the data, the cursor will be closed when you run commands and since autocommit will be disabled (False by default), the transaction will be rolled back. You seem to have full control over what you commit inside the operator you use.

But indeed each "operator" is necessarily separate transaction - there are no transaction that spans multiple operators (and cannot be because each task is run in a separate process or even in a separate machine). So keeping transaction open and rolling it back in another operator is indeed not possible. And in standard PostgresOperator, complex logic is not really possible to 

However what you are really after in this case is not to use pre-existing operator, but write your own that performs more complex logic. And you can easily do that with @task decorator - and you can look at the PostgresOperator implementation to see how you can implement 
Smth like that should work:

```
@task
def my_task():
     hook = PostgresHook(....)
     with closing(self.get_conn()) as conn):
         cur = conn.cursor()
         cur.execute("SOME SQL TO UPDATE", parameters)
         res = cur.execute("SOME CHECK SQL")
         if res....: # condition to commit
           res.commit()
```

**Reply by volvocarsmatt** [reply]
Thanks for confirming! Since SQLCheckOperator is is its own operator we could and least not use that want if we want this functionality. I learned that this is not the behaviour of the popular dbt either.

As long as we don't have autocommit on every statement on the PostgresOperator, we should still be able to rollback if an error occurs  in there? If we write the test in SQL that raise a SQL error, this would give us the rollback feature. However, we can use the provided SQLCheckOperator for this.

Let's see if I can write an example that proves this, but something like

```sql
INSERT INTO table1 (field1) VALUES ('my_error')
DO $$ BEGIN ASSERT (SELECT COUNT(field1) FROM table1 WHERE field1 = 'my_error') = 0, 'not 0!';END;$$;
```

One alternative strategy is to run your tests in a test environment and hope it reflects production.

**Reply by volvocarsmatt** [reply]
It is indeed working

```python
import datetime

from airflow import DAG
from airflow.providers.postgres.operators.postgres import PostgresOperator

DAG_ID = "rollback_task_dag"

with DAG(
        dag_id=DAG_ID,
        start_date=datetime.datetime(2022, 10, 2),
        schedule_interval="@once",
        catchup=False,
) as dag:
    recreate_colour_table = PostgresOperator(
        task_id="recreate_rainbow_colour_table",
        sql="""
            DROP TABLE IF EXISTS colours_of_the_rainbow;
            CREATE TABLE colours_of_the_rainbow (
            name VARCHAR PRIMARY KEY,
            animal VARCHAR NOT NULL
            );
            
            INSERT INTO colours_of_the_rainbow (name, animal)
            VALUES ('red', 'fox');
            INSERT INTO colours_of_the_rainbow (name, animal)
            VALUES ('yellow', 'lion');
        """
    )

    failing_task = PostgresOperator(
        task_id="built_in_test",
        sql="""
            INSERT INTO colours_of_the_rainbow (name, animal)
            VALUES ('teal', 'dragonfly');

            DO $$ BEGIN ASSERT (SELECT 1 FROM colours_of_the_rainbow WHERE name = 'teal') IS NULL, 'teal is not in rainbow';END;$$;
        """
    )

    recreate_colour_table >> failing_task
```

Creates the log 

```console
psycopg2.errors.AssertFailure: teal is not in rainbow
CONTEXT:  PL/pgSQL function inline_code_block line 1 at ASSERT
```

And the previous transaction is saved as expected. Not the cleanest approach, but maybe you can write a UDF
